src/masternode-sync.cpp

```diff
@@ -274,13 +274,18 @@ void CMasternodeSync::Process()
     if (tick++ % MASTERNODE_SYNC_TIMEOUT != 0) return;
 
     if (IsSynced()) {
-        /*
-            Resync if we lose all masternodes from sleep/wake or failure to sync originally
-        */
-        if (mnodeman.CountEnabled(true /* only_legacy */) == 0 && !isRegTestNet) {
+        if (isRegTestNet) {
+            return;
+        }
+        bool legacy_obsolete = deterministicMNManager->LegacyMNObsolete();
+        // Check if we lost all masternodes from sleep/wake or failure to sync originally (after
+        // spork 21, check if we lost all proposals instead). If we did, resync from scratch.
+        if ((!legacy_obsolete && mnodeman.CountEnabled(true /* only_legacy */) == 0) ||
+            (legacy_obsolete && g_budgetman.CountProposals() == 0)) {
```

Comment:
Better to skip testnet here. The proposals count is often 0, so this will cause the node to reset the sync process for no reason.
(this was going to happen with the previous code as well, previous `CountEnabled` after DIP3 enforcement was going to return 0 as well).

Reply:
> this was going to happen with the previous code as well

Yep, that's why I added this commit. So, at least, main net mnsync is not reset.
I thought about the testnet issue, but I would rather not skip it (to be able to test mainnet-like scenarios).
Better to discuss a different approach on another PR (if there is no proposal on testnet, that means that, after spork 21, there is nothing to sync anyway, besides sporks).